ex/app/src/main/java/com/example/ex/RecyclerViewAdapter.java

```diff
@@ -0,0 +1,113 @@
+package com.example.ex;
+
+import android.app.Activity;
+import android.util.Log;
+import android.view.LayoutInflater;
+import android.view.View;
+import android.view.ViewGroup;
+import android.widget.CheckBox;
+import android.widget.RatingBar;
+import android.widget.TextView;
+import androidx.annotation.NonNull;
+import androidx.recyclerview.widget.RecyclerView;
+import java.util.ArrayList;
+
+public class RecyclerViewAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {
+
+    private Activity context;
+    private ArrayList<Tuple> tupleArrayList;
+    private boolean flag = true;
+
+    RecyclerViewAdapter(final Activity context, final ArrayList<Tuple> tupleArrayList) {
+        this.context = context;
+        this.tupleArrayList = tupleArrayList;
+    }
+
+    ArrayList<Tuple> getTupleArrayList(){
+        return tupleArrayList;
+    }
+
+    @NonNull
+    @Override
+    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull final ViewGroup parent, final int viewType) {
+        final View rootView;
+        if (flag)
+        {
+            flag = false;
+            rootView = LayoutInflater.from(context)
+                    .inflate(R.layout.adapter_item_with_custom_ratingbar,parent,false);
+        } else {
+            rootView = LayoutInflater.from(context)
+                    .inflate(R.layout.adapter_item_with_star_ratingbar,parent,false);
+        }
+
+        return new RecyclerViewViewHolder(rootView);
+    }
+
+    @Override
+    public void onBindViewHolder(@NonNull final RecyclerView.ViewHolder holder, final int position) {
+        final Tuple tuple = tupleArrayList.get(position);
+        final RecyclerViewViewHolder viewHolder = (RecyclerViewViewHolder) holder;
+        viewHolder.txtView_title.setText(tuple.getTitle());
+
+        setRating(tuple, viewHolder);
+        setCheckBox(tuple, viewHolder);
+    }
+
+    private void setRating(final Tuple tuple, final RecyclerViewViewHolder viewHolder){
+        viewHolder.ratingBar.setRating(tuple.getRating());
+        viewHolder.ratingBar.setOnRatingBarChangeListener(new RatingBar.OnRatingBarChangeListener() {
+            @Override
+            public void onRatingChanged(final RatingBar ratingBar,
+                                        final float rating, final boolean fromUser) {
+                tuple.setRating((int)rating);
```

Comment:
One of the conditions of the task was that cell classes should be immutable. 

In this scheme the flow looks like this:
1. `ViewModel` dispatches some initial data to the `Fragment` via `LiveData`
1. User performs some action, our `ViewModel` is notified about that via some `onSomethingChanged` method
1. `ViewModel` updates it's internal state, rebuilds the list of cells and other view state classes and dispatches it via `LiveData`

In this scheme some mutable data can be inside the `ViewModel`, but our `Fragment` simply receives immutable view state / cell classes and renders them.
Let's adopt this scheme here.

Also we expect that the `Adapter` will be implemented in accordance with our `RecyclerView` guide (`Adapter with several view types` section, you can use [`ListAdapter`](https://developer.android.com/reference/androidx/recyclerview/widget/ListAdapter) instead of our `AbsDiffableListAdapter` here). The part of guides about emulating `sealed class` in Java will also be useful.